File: src/main/java/com/jigumulmi/admin/banner/dto/response/AdminBannerPlaceListResponseDto.java
package com.jigumulmi.admin.banner.dto.response;

import com.jigumulmi.admin.banner.dto.response.AdminBannerPlaceListResponseDto.BannerPlaceDto;
import com.jigumulmi.common.PagedResponseDto;
import com.jigumulmi.place.domain.Place;
import com.jigumulmi.place.dto.response.PlaceCategoryDto;
import com.jigumulmi.place.dto.response.SubwayStationResponseDto;
import com.jigumulmi.place.vo.District;
import java.util.List;
import lombok.Builder;
import lombok.Getter;
import lombok.experimental.SuperBuilder;

@Getter
@SuperBuilder
public class AdminBannerPlaceListResponseDto extends PagedResponseDto<BannerPlaceDto> {

    @Getter
    @Builder
    public static class BannerPlaceDto {

        private Long id;
        private String name;
        private District district;
        private SubwayStationResponseDto subwayStation;
        private List<PlaceCategoryDto> categoryList;

        public static BannerPlaceDto from(Place place) {
            SubwayStationResponseDto subwayStationResponseDto;
            if (place.getSubwayStationPlaceList().isEmpty()) {
                subwayStationResponseDto = null;
            } else {
                subwayStationResponseDto = SubwayStationResponseDto.fromMainStation(
                    place.getSubwayStationPlaceList().getFirst().getSubwayStation()
                );
            }

            return BannerPlaceDto.builder()
                .id(place.getId())
                .name(place.getName())
                .district(place.getDistrict())
                .subwayStation(subwayStationResponseDto)
                .categoryList(place.getCategoryMappingList().stream()
                    .map(PlaceCategoryDto::fromPlaceCategoryMapping).toList())
                .build();
        }
    }
}

File: admin/src/main/kotlin/com/jigumulmi/banner/dto/response/AdminBannerPlaceListResponseDto.kt
package com.jigumulmi.banner.dto.response

import com.jigumulmi.common.AdminPagedResponseDto
import com.jigumulmi.place.domain.Place
import com.jigumulmi.place.dto.response.PlaceCategoryDto
import com.jigumulmi.place.dto.response.SubwayStationResponseDto
import com.jigumulmi.place.vo.District

data class AdminBannerPlaceListResponseDto(
    override val page: PageDto,
    override val data: List<BannerPlaceDto>
) : AdminPagedResponseDto<AdminBannerPlaceListResponseDto.BannerPlaceDto>(page, data) {

    data class BannerPlaceDto(
        val id: Long,
        val name: String? = null,
        val district: District? = null,
        val subwayStation: SubwayStationResponseDto? = null,
        val categoryList: List<PlaceCategoryDto> = emptyList()
    ) {
        companion object {
            fun from(place: Place): BannerPlaceDto {
                val subwayStationResponseDto = if (place.subwayStationPlaceList.isEmpty()) {
                    null
                } else {
                    SubwayStationResponseDto.fromMainStation(
                        place.subwayStationPlaceList.first().subwayStation
                    )
                }

                return BannerPlaceDto(
                    id = place.id,
                    name = place.name,
                    district = place.district,
                    subwayStation = subwayStationResponseDto,
                    categoryList = place.categoryMappingList.map { mapping ->
                        PlaceCategoryDto.fromPlaceCategoryMapping(mapping)
                    }
                )
            }
        }
    }
}
